From c0fc54d4a546c330dee5e3e76756a68b67b42c1e Mon Sep 17 00:00:00 2001
From: Tommy Webb <tommy@calyxinstitute.org>
Date: Wed, 29 Mar 2023 14:39:16 -0400
Subject: [PATCH] Restrict USB at boot time in "unlocked" mode

When Restrict USB is set to allow "when unlocked", USB devices that are
already connected at boot time should be ejected immediately, since the
device has yet to be unlocked.

Issue: calyxos#1414
Change-Id: Iac537b4dc035761a4a9d640f2ce3434a94be3c90
---
 .../java/com/android/server/usb/UsbDeviceManager.java    | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/services/usb/java/com/android/server/usb/UsbDeviceManager.java b/services/usb/java/com/android/server/usb/UsbDeviceManager.java
index 162c95f87ad1..f20a9d33904f 100644
--- a/services/usb/java/com/android/server/usb/UsbDeviceManager.java
+++ b/services/usb/java/com/android/server/usb/UsbDeviceManager.java
@@ -1906,11 +1906,12 @@ public class UsbDeviceManager implements ActivityTaskManagerInternal.ScreenObser
             final int restrictUsb = LineageSettings.Global.getInt(mContentResolver,
                     LineageSettings.Global.TRUST_RESTRICT_USB, 1);
             // Effective immediately, ejects any connected USB devices.
-            // If the restriction is set to "only when locked", only execute once USB is
-            // disconnected and keyguard is showing, to avoid ejecting connected devices
-            // on lock
+            // If the restriction is set to "allow when unlocked", only execute once USB is
+            // disconnected and keyguard is showing, to avoid ejecting connected devices on lock,
+            // unless the device has not finished booting, in which case it has never unlocked.
             final boolean usbConnected = mConnected || mHostConnected;
-            final boolean shouldRestrict = (restrictUsb == 1 && mIsKeyguardShowing && !usbConnected)
+            final boolean shouldRestrict =
+                    (restrictUsb == 1 && mIsKeyguardShowing && (!mBootCompleted || !usbConnected))
                     || restrictUsb == 2;
 
             UsbManager usbManager = mContext.getSystemService(UsbManager.class);
-- 
GitLab

